presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotPlanOptimizer.java

```diff
@@ -194,6 +202,13 @@ public Visitor(TableScanNode tableScanNode, ConnectorSession session, PlanNodeId
         @Override
         public PlanNode visitPlan(PlanNode node, Void context)
         {
+            if (node instanceof SetOperationNode) {
+                final ImmutableList<PlanNode> newChildren = node.getSources()
+                        .stream()
+                        .map(source -> visitPlan(source, context))
+                        .collect(toImmutableList());
+                return node.replaceChildren(newChildren);
+            }
             Optional<PlanNode> pushedDownPlan = tryCreatingNewScanNode(node);
```

Comment:
This recursion looks weird. `Visitor` itself is recursive. So ideally, we won't need this extra recursion.  IMO, we probably need to fix this `visitPlan` call so that it will naturally do recursion not only to set operations but all other plan nodes.

Reply:
Make sense, I moved this logic inside the `Visitor` itself.
The old Visitor expects only one `TableScanNode` and sets it before construct `Visitor`. Now I make it to a context to carry, so SetOperationNode with multiple `TableScanNode` will just visit all source nodes.